[Serializer] Added a ConstraintViolationListNormalizer

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #11309
| License       | MIT
| Doc PR        | -

---

It seems logical to me that Symfony is able to serialise natively some very common Symfony data structure. (and requested by @nicolas-grekas & @javiereguiluz )

Usage example (from symfony/symfony-demo):

```php
    /**
     * @Route("", name="api_blog_new")
     * @Method("POST")
     * @Security("is_granted('ROLE_ADMIN')")
     */
    public function newAction(Request $request)
    {
        $data = $request->getContent();

        $post = $this->get('serializer')->deserialize($data, Post::class, 'json', ['groups' => ['post_write']]);

        $post->setAuthor($this->getUser());

        $violations = $this->get('validator')->validate($post);

        $post->setSlug($this->get('slugger')->slugify($post->getTitle()));

        if (count($violations) > 0) {
            $repr = $this->get('serializer')->serialize($violations, 'json');

            return JsonResponse::fromJsonString($repr, 400);
        }

        $this->getDoctrine()->getManager()->persist($post);
        $this->getDoctrine()->getManager()->flush();

        $repr = $this->get('serializer')->serialize($post, 'json', ['groups' => ['post_read']]);

        return JsonResponse::fromJsonString($repr);
    }
```

lyrixx committed:
[Serializer] Added a ConstraintViolationListNormalizer

theofidry commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1202,6 +1203,13 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
             $definition->addTag('serializer.normalizer', array('priority' => -900));
         }
 
+        if (class_exists('Symfony\Component\Serializer\Normalizer\ConstraintViolationListNormalizer')) {
```
`ConstraintViolationListNormalize::class`

lyrixx replied:
No, I want to be consistent with other usage in this file. just above and bellow

ro0NL replied:
Had the same issue with https://github.com/symfony/symfony/pull/20567/files#diff-1cf8b3cc93de9c5674101413d27a9bf0R47, at first i thought; lets be consistent. On 2nd thought i chose to _just_ use `::class` notation for new code, as it's the better syntax en exposes class deps (ie. it improves usage tracking, and therefor maintenance).

javiereguiluz commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,53 @@
+
+        return array(
+            'title' => 'An error occurred',
+            'detail' => $messages ? implode("\n", $messages) : (string) $violationsList,
```
I'm not sure about Symfony's policy about this, but do we use `PHP_EOL` instead of `\n`?

lyrixx replied:
Updated

javiereguiluz reviewed (approved):
👍  it looks like a useful feature to me! Thanks @lyrixx.

ro0NL commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        return array(
+            'title' => 'An error occurred',
```
Isnt this format rather opinionated? Do we really need it compared to `return $violations;`.

Point is; i like to integrate such a normalizer in one of my api response providers, by doing something like;

```
return new JsonResponse(['our_api_format' => ['errors' => $normalizer->normalize($violations)]]);
```

Any thoughts?

lyrixx replied:
You are right. Actually I adapted code from API platform (that's why I kept Kevin as author)
But actually I like this format. But I'm open to suggestions.

dunglas commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,53 @@
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ConstraintViolationListNormalizer implements NormalizerInterface
```
The description looks not accurate.

lyrixx replied:
Fixed.

dunglas commented:
The original implementation (https://github.com/api-platform/core/blob/master/src/Hydra/Serializer/ConstraintViolationListNormalizer.php) was following the Hydra W3C draft.

WDYT about keeping Hydra support in Symfony?

dunglas commented:
Another alternative is to implement RFC7807 (supported by API Platform and Zend Apigility): https://github.com/api-platform/core/blob/master/src/Problem/Serializer/ConstraintViolationListNormalizer.php

javiereguiluz commented:
RFC7807 looks like the ideal solution to create something standard ... but its status is still: "PROPOSED STANDARD". Can we guess the chances of being changed a lot in the future or even rejected?

lyrixx commented:
I implemented RFC7807 (supported by API Platform and Zend Apigility)

fabpot commented:
Can you add a note about the fact that it implements RFC7807 in the phpdocs?

stof commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,53 @@
+        foreach ($object as $violation) {
+            $violations[] = array(
+                'propertyPath' => $violation->getPropertyPath(),
+                'message' => $violation->getMessage(),
```
I would also add the code (`$violation->getCode()`) for the machine-readable stable identifier of the violation (the message is not stable as it is translated)

lyrixx replied:
Good idea.

lyrixx commented:
Comment addressed. PR updated.

Note: I also renamed `propertyPath` to property_path`.

ro0NL commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,56 @@
+        return array(
+            'type' => isset($context['type']) ? $context['type'] : 'https://tools.ietf.org/html/rfc2616#section-10',
+            'title' => isset($context['title']) ? $context['title'] : 'An error occurred',
+            'detail' => $messages ? implode("\n", $messages) : (string) $object,
```
Not sure we can safely rely on `(string) $object` (`__toString` is not part of the interface). Do we really need the fallback?

lyrixx replied:
Removed.

dunglas commented:
> Note: I also renamed propertyPath to `property_path`

Why this change? AFAIK, camel case is preferred (for instance, it's what Google and Schema.org use).

lyrixx commented:
Ah ok. I will revert it so. Usually I prefer Snake Case in my array / API...

lyrixx commented:
Reverted. The PR is now ready.

dunglas commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,56 @@
+        );
+    }
+
+    public function supportsNormalization($data, $format = null)
```
`{@inheritdoc}` is missing.

lyrixx replied:
Thanks. Fixed

dunglas commented:
👍 for me. @teohhanhui would you mean reviewing this PR? I know you studied this RFC.

teohhanhui commented:
This PR seems to bring over the same mistakes as I've pointed out in api-platform/core#793

Reproducing it here:

The current implementation is not RFC 7807 compliant:

> Consumers MUST use the "type" string as the primary identifier for the problem type
> (https://tools.ietf.org/html/rfc7807#section-3.1)
> 
> A problem's type URI SHOULD resolve to HTML documentation that explains how to resolve the problem.
> (https://tools.ietf.org/html/rfc7807#section-4)

But instead we're using a link to the "Status Code Definitions" section of RFC 2616 (deprecated and replaced by RFC 7231).

So I propose we should omit the "type" altogether, which is allowed:

> When this member is not present, its value is assumed to be "about:blank".
> (https://tools.ietf.org/html/rfc7807#section-3.1)

In which case the spec further states:

> When "about:blank" is used, the title SHOULD be the same as the recommended HTTP status phrase for that code (e.g., "Not Found" for 404, and so on)
> (https://tools.ietf.org/html/rfc7807#section-4.2)

teohhanhui commented:
Since we're normalizing Symfony Validator constraint violations, it'd be perfect if we have specific pages on symfony.com to explain each violation (this would be what we put in the `type` property of the Problem Details object). We already have a UUID for each violation, so let's make full use of that.

dunglas commented:
Any update about this one?

lyrixx commented:
What should I do ?

dunglas commented:
Fixing the comments raised by @teohanhui

lyrixx commented:
I have rebased this PR and now it's targeting 4.1 (I guess it's too late for 3.4 ?)

I have removed the `type` attribute of the response.

This normalizer still does not really implement the RFC. It's not really possible. We can not get the HTTP Status code of the response.

So I don't know how to unlock this PR. May be we can simply remove the reference to the RFC, or to explain it does not follow each rules of the RFC.

What do you think ?

dunglas commented:
The status code can be passed through the context (it can be optional).

lyrixx commented:
> The status code can be passed through the context (it can be optional).

But it's not mandatory, and it's boring if you need to pass 4XX each time ... :/

lyrixx commented:
I have rebased the PR

lyrixx commented:
Ping

fabpot commented:
@lyrixx "it's not ready" -> I suppose you meant "it's now ready", right?

fabpot commented on src/Symfony/Component/Serializer/CHANGELOG.md:
```diff
@@ -4,11 +4,11 @@ CHANGELOG
+   of objects that needs data insertion in constructor
+ * added an optional `default_constructor_arguments` option of context to specify a default data in
+   case the object is not initializable by its constructor because of data missing
+ * added optional `bool $escapeFormulas = false` argument to `CsvEncoder::__construct`
```
Looks like you haven't added anything here.

fabpot commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,58 @@
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+
+/**
+ * A normalizer that normalize a ConstraintViolationListInterface instance.
```
normalizes

fabpot commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * A normalizer that normalize a ConstraintViolationListInterface instance.
+ *
+ * This Normalizer implements RFC7807 {@link https://tools.ietf.org/html/rfc7807}.
```
If we're not fully compliant, it should be added here.

lyrixx commented:
> @lyrixx "it's not ready" -> I suppose you meant "it's now ready", right?

Oups, You are right.

---

I have addressed your comment. It should be OK now

fabpot commented on src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * A normalizer that normalizes a ConstraintViolationListInterface instance.
+ *
+ * This Normalizer implements partially RFC7807 {@link https://tools.ietf.org/html/rfc7807}.
```
I would say `implements RFC7807 partially.`

But can we be a bit more precise? Can we list what we do not support or how we diverge from the RFC? I think that would help our users.

lyrixx replied:
Actually I re-read the RFC and came up to the same conclusion as @teohhanhui's [comment](https://github.com/symfony/symfony/pull/22150#issuecomment-291249290).

We are allowed to omit the `type` property.

That's why I removed the "partially" because the normalizer do implement the RFC.

fabpot reviewed (approved):
minor comment

fabpot commented:
Thank you @lyrixx.

teohhanhui commented on src/Symfony/Component/Serializer/Tests/Normalizer/ConstraintViolationListNormalizerTest.php:
```diff
@@ -0,0 +1,65 @@
+        ));
+
+        $expected = array(
+            'title' => 'An error occurred',
```
RFC 7807 is unambiguous on this point:

> **[3.1](https://tools.ietf.org/html/rfc7807#section-3.1).  Members of a Problem Details Object**
>
>    A problem details object can have the following members:
>
>    - "type" (string) - ... When
>    this member is not present, its value is assumed to be
>    "about:blank".

> [4.2](https://tools.ietf.org/html/rfc7807#section-4.2).  Predefined Problem Types
>
>    This specification reserves the use of one URI as a problem type:
>
>    The "about:blank" URI [RFC6694], when used as a problem type,
>    indicates that the problem has no additional semantics beyond that of
>    the HTTP status code.
>
>    When "about:blank" is used, the title SHOULD be the same as the
>    recommended HTTP status phrase for that code (e.g., "Not Found" for
>    404, and so on), although it MAY be localized to suit client
>    preferences (expressed with the Accept-Language request header).
>
>    Please note that according to how the "type" member is defined
>    (Section 3.1), the "about:blank" URI is the default value for that
>    member.  Consequently, any problem details object not carrying an
>    explicit "type" member implicitly uses this URI.

Perhaps:

```json
{
  "type": "/validation-error",
  "title": "Validation failed",
  ...
}
```
